src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
+ */
+bool
+supervisor_add_dynamic(Supervisor *supervisor, Service *service)
+{
+	Service *dynamicService;
+	sigset_t sig_mask;
+	sigset_t sig_mask_orig;
+	int serviceIndex = supervisor->dynamicServices.serviceCount;
+	bool started;
+
+	if (supervisor_service_exists(supervisor, service->name))
+	{
+		log_error("Service %s already exists under supervision",
+				  service->name);
+		return false;
+	}
+
+	/* 0-based */
+	if (serviceIndex >= (MAXDYNSERVICES - 1))
+	{
+		log_error("Reached maximum permitted number of dynamic services, "
+				  "service %s is not added",
+				  service->name);
+		return false;
+	}
+
+	/*
+	 * Block signals until the service has been added in the dynamic array
+	 * unless failed to start. Failing to block signals, opens a race window.
+	 */
+	if (!block_signals(&sig_mask, &sig_mask_orig))
+	{
+		return false;
+	}
+
+	/*
+	 * In case there is an active signal to stop, set prior to blocking signals,
+	 * do not add the service. Our caller should be able to handle this case.
+	 */
+	if (asked_to_stop || asked_to_stop_fast || asked_to_reload || asked_to_quit)
+	{
+		/* restore signal masks (un block them) now */
+		(void) unblock_signals(&sig_mask_orig);
+		return false;
+	}
+
+	started = (*service->startFunction)(service->context, &(service->pid));
+
+	if (started)
+	{
+		uint64_t now = time(NULL);
+		RestartCounters *counters = &(service->restartCounters);
+
+		counters->count = 1;
+		counters->position = 0;
+		counters->startTime[counters->position] = now;
+
+		log_info("Started pg_autoctl %s service with pid %d",
+				 service->name, service->pid);
+	}
+	else
+	{
+		log_error("Failed to start pg_autoctl %s service",
+				  service->name);
+		(void) unblock_signals(&sig_mask_orig);
+		return false;
+	}
+
+	/*
+	 * IGNORE-BANNED fits here because we want to take ownership of the service.
+	 * We copy to an address that we own from a non overlapping address.
+	 */
+	dynamicService = &supervisor->dynamicServices.services[serviceIndex];
+	memcpy(dynamicService, service, sizeof(*service)); /* IGNORE-BANNED */
+	supervisor->dynamicServices.serviceCount++;
+
+	(void) unblock_signals(&sig_mask_orig);
+
+	return true;
+}
+
+
+/*
+ * supervisor_remove_dynamic stops a running dynamic service with serviceName.
+ * Once the service is stopped, then it is removed from the dynamicService array
+ * and supervisor_loop can largely forget about it.
+ *
+ * However, it will get notified during waitpid, so we add this service to the
+ * recentlyRemoved array and we will hold it there until it is requested via
+ * supervisor_find_recently_removed_dynamic() which subsequently removes it.
+ *
+ * Returns true if the service is stopped and removed from the dynamicService
+ * array, false otherwise.
+ */
+bool
+supervisor_remove_dynamic(Supervisor *supervisor, const char *serviceName)
+{
+	Service *dynamicService = NULL;
+	int serviceCount = supervisor->dynamicServices.serviceCount;
+	int serviceIndex;
+
+	/* Find it in the dynamic array */
```

Comment:
Could we use the `supervisor_service_exists` function if it had a better API for this call-site? And then I suppose it would be renamed `supervisor_dynamic_service_find` or something.

Reply:
Those two things are solving a very different problem, or at least in my mind they do :)

supervisor_service_exists() is intended to find any active service under supervision. That is what is exposed to our callers. I do not want the callers of the supervisor to have more access to it.

Privately in supervisor.c we can add as many helper functions as desired.
What do you think?

Reply:
Oh I meant just about the loop over the services within the `supervisor_remove_dynamic` function itself. Not the whole thing, just that part that finds the service in the array.